File: c/benchmark.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <math.h>
#include <stdbool.h>

// assinatura das funções
bool subset_sum_backtracking(int conjunto[], int n, int alvo);
bool subset_sum_dp(int conjunto[], int n, int alvo);

// função auxiliar para calcular o tempo em milissegundos
double calcular_tempo_ms(struct timespec start, struct timespec end) {
    return (end.tv_sec - start.tv_sec) * 1000.0 + (end.tv_nsec - start.tv_nsec) / 1000000.0;
}

void testar_e_salvar(const char* algoritmo, const char* cenario, int n, int conjunto[], int alvo, FILE* csv, bool eh_dp){

    int NUM_RODADAS = 30;
    double tempos_rodadas[NUM_RODADAS];
    double soma_tempos = 0;

    // REMOVIDO o loop for (int t = 0...), mantendo apenas o loop das 30 rodadas
    for (int rodada = 0; rodada < NUM_RODADAS; rodada++) {
        struct timespec start, end;
        
        clock_gettime(CLOCK_MONOTONIC, &start);
        
        if (eh_dp) {
            subset_sum_dp(conjunto, n, alvo);
        } else {
            subset_sum_backtracking(conjunto, n, alvo);
        }
        
        clock_gettime(CLOCK_MONOTONIC, &end);
        
        double tempo_gasto = calcular_tempo_ms(start, end);
        tempos_rodadas[rodada] = tempo_gasto;
        soma_tempos += tempo_gasto;
    }

    // cálculos estatísticos
    double media = soma_tempos / NUM_RODADAS;
    double variancia = 0;

    for (int r = 0; r < NUM_RODADAS; r++){
        variancia += pow(tempos_rodadas[r] - media, 2);
    }

    variancia /= NUM_RODADAS;
    double desvio_padrao = sqrt(variancia);

    fprintf(csv, "%s,%s,%d,%.4f,%.4f\n", algoritmo, cenario, n, media, desvio_padrao);
    
    fflush(csv); 

}

int main(){

    srand(time(NULL));

    int tamanhos[] = {10, 20, 30};
    int num_tamanhos = sizeof(tamanhos) / sizeof(tamanhos[0]);

    FILE *csv = fopen("../dados/resultados_c.csv", "w");
    if (csv == NULL) {
        printf("Erro: Crie uma pasta chamada 'dados' no diretorio anterior!\n");
        return 1;
    }
    fprintf(csv, "Algoritmo, Cenario, Tamanho_N, Tempo_Medio_ms, Desvio_Padrao_ms\n");

    FILE *entradas_py = fopen("../dados/entradas_compartilhadas.txt", "w");

    printf("Iniciando bateria de testes. Isso pode demorar alguns minutos...\n");

    for (int t = 0; t < num_tamanhos; t++) {

        int n = tamanhos[t];

        int *vetor_melhor = (int*)malloc(n * sizeof(int));
        int *vetor_medio = (int*)malloc(n * sizeof(int));
        int *vetor_pior = (int*)malloc(n * sizeof(int));

        // 1. GERANDO MELHOR CASO
        vetor_melhor[0] = 150;
        int alvo_melhor = 150;
        for (int i = 1; i < n; i++) vetor_melhor[i] = rand() % 100 + 1;

        // 2. GERANDO CASO MÉDIO
        int soma_parcial = 0;
        for (int i = 0; i < n; i++) {
            vetor_medio[i] = rand() % 100 + 1;
            if (i % 3 == 0) soma_parcial += vetor_medio[i];
        }
        int alvo_medio = soma_parcial;  // alvo existe mas requer busca

        // 3. GERANDO PIOR CASO
        int soma_total = 0;
        for (int i = 0; i < n; i++) {
            vetor_pior[i] = (rand() % 50 + 1) * 2;
            soma_total += vetor_pior[i];
        }
        int alvo_pior = soma_total + 1; // ímpar inatingível

        // EXPORTANDO AS ENTRADAS PARA PYTHON LER
        fprintf(entradas_py, "TAMANHO:%d\n", n);
        
        fprintf(entradas_py, "ALVO_MELHOR:%d\nVETOR_MELHOR:", alvo_melhor);
        for(int i=0; i<n; i++) fprintf(entradas_py, "%d,", vetor_melhor[i]);
        
        fprintf(entradas_py, "\nALVO_MEDIO:%d\nVETOR_MEDIO:", alvo_medio);
        for(int i=0; i<n; i++) fprintf(entradas_py, "%d,", vetor_medio[i]);
        
        fprintf(entradas_py, "\nALVO_PIOR:%d\nVETOR_PIOR:", alvo_pior);
        for(int i=0; i<n; i++) fprintf(entradas_py, "%d,", vetor_pior[i]);
        
        fprintf(entradas_py, "\n---\n");

        fflush(entradas_py); 

        // EXECUTANDO OS TESTES (Backtracking)
        testar_e_salvar("Backtracking", "Melhor Caso", n, vetor_melhor, alvo_melhor, csv, false);
        testar_e_salvar("Backtracking", "Caso Medio", n, vetor_medio, alvo_medio, csv, false);
        testar_e_salvar("Backtracking", "Pior Caso", n, vetor_pior, alvo_pior, csv, false);

        // EXECUTANDO OS TESTES (Programação Dinâmica)
        testar_e_salvar("DP", "Melhor Caso", n, vetor_melhor, alvo_melhor, csv, true);
        testar_e_salvar("DP", "Caso Medio", n, vetor_medio, alvo_medio, csv, true);
        testar_e_salvar("DP", "Pior Caso", n, vetor_pior, alvo_pior, csv, true);

        // Liberando memória
        free(vetor_melhor);
        free(vetor_medio);
        free(vetor_pior);
        
        printf("Concluido tamanho n=%d...\n", n);
    }

    fclose(csv);
    fclose(entradas_py);
    printf("Todos os testes finalizados! Resultados em dados/resultados_c.csv\n");

    return 0;
}

File: python/benchmark.py
"""
benchmark.py — Geração de entradas e coleta de resultados para Subset Sum (Python).
 
Metodologia:
  - Backtracking: tamanhos pequenos (10, 15, 20) — complexidade O(2^n) explode rapidamente
  - DP:           tamanhos maiores (50, 200, 500) — complexidade O(n×W), escala bem
  - 30 rodadas por (algoritmo × cenário × tamanho)
  - Calcula média e desvio-padrão do tempo em milissegundos
  - Salva os resultados em dados/resultados_python.csv
 
Precisão de tempo: time.perf_counter() — alta resolução, equivalente a
clock_gettime(CLOCK_MONOTONIC) no C.
"""
 
import time
import math
import csv
import os
import sys
import random
 
sys.setrecursionlimit(200_000)
 
from backtracking import subset_sum_backtracking
from dp import subset_sum_dp
 
 
# ---------------------------------------------------------------------------
# Configuração — tamanhos separados por algoritmo
# ---------------------------------------------------------------------------
 
NUM_RODADAS        = 30
TAMANHOS_BT        = [10, 15, 20]    # backtracking: O(2^n), limitar para não travar
TAMANHOS_DP        = [50, 100, 200]  # DP: O(n×W), aguenta entradas bem maiores
DADOS_DIR          = os.path.join(os.path.dirname(__file__), "..", "dados")
ENTRADAS_PATH      = os.path.join(DADOS_DIR, "entradas_compartilhadas.txt")
SAIDA_CSV          = os.path.join(DADOS_DIR, "resultados_python.csv")
 
 
# ---------------------------------------------------------------------------
# Auxiliares
# ---------------------------------------------------------------------------
 
def calcular_tempo_ms(func, *args) -> float:
    inicio = time.perf_counter()
    func(*args)
    fim = time.perf_counter()
    return (fim - inicio) * 1000.0
 
 
def estatisticas(tempos: list[float]) -> tuple[float, float]:
    media = sum(tempos) / len(tempos)
    variancia = sum((t - media) ** 2 for t in tempos) / len(tempos)
    return media, math.sqrt(variancia)
 
 
def testar_e_salvar(algoritmo: str, cenario: str, n: int,
                    conjunto: list[int], alvo: int,
                    writer, eh_dp: bool) -> None:
    func = subset_sum_dp if eh_dp else subset_sum_backtracking
    tempos = [calcular_tempo_ms(func, conjunto, n, alvo) for _ in range(NUM_RODADAS)]
    media, desvio = estatisticas(tempos)
    writer.writerow([algoritmo, cenario, n, f"{media:.4f}", f"{desvio:.4f}"])
    print(f"  [{algoritmo:12s}] {cenario:12s} n={n:3d} → "
          f"média={media:.4f} ms  dp={desvio:.4f} ms")
 
 
# ---------------------------------------------------------------------------
# Geração de entradas por tamanho e alvo fixo
# ---------------------------------------------------------------------------
 
def gerar_bloco(n: int, alvo_fixo: int | None = None) -> dict:
    """
    Gera os três cenários para um dado n.
    alvo_fixo: se informado, usa esse valor como alvo_melhor (para compatibilidade
               com entradas_compartilhadas.txt do C, onde alvo_melhor = 150).
    """
    # Melhor caso: alvo está no primeiro índice — backtracking/DP termina cedo
    alvo_melhor = alvo_fixo if alvo_fixo else 150
    vetor_melhor = [alvo_melhor] + [random.randint(1, 100) for _ in range(n - 1)]
 
    # Caso médio: alvo existe mas requer busca
    vetor_medio = [random.randint(1, 100) for _ in range(n)]
    alvo_medio = sum(vetor_medio[i] for i in range(0, n, 3))
 
    # Pior caso: todos pares, alvo ímpar inatingível
    vetor_pior = [random.randint(1, 50) * 2 for _ in range(n)]
    alvo_pior = sum(vetor_pior) + 1
 
    return {
        "n": n,
        "melhor": (vetor_melhor, alvo_melhor),
        "medio":  (vetor_medio,  alvo_medio),
        "pior":   (vetor_pior,   alvo_pior),
    }
 
 
# ---------------------------------------------------------------------------
# Parser do arquivo de entradas compartilhadas (gerado pelo C)
# ---------------------------------------------------------------------------
 
def parse_entradas(path: str) -> dict[int, dict]:
    """Lê entradas_compartilhadas.txt e retorna um dict keyed por n."""
    resultado = {}
    with open(path, "r") as f:
        conteudo = f.read()
    for bloco in conteudo.split("---"):
        bloco = bloco.strip()
        if not bloco:
            continue
        linhas = {l.split(":")[0]: l.split(":", 1)[1] for l in bloco.splitlines() if ":" in l}
        n           = int(linhas["TAMANHO"])
        alvo_melhor = int(linhas["ALVO_MELHOR"])
        vetor_melhor = [int(x) for x in linhas["VETOR_MELHOR"].split(",") if x.strip()]
        alvo_medio  = int(linhas["ALVO_MEDIO"])
        vetor_medio  = [int(x) for x in linhas["VETOR_MEDIO"].split(",") if x.strip()]
        alvo_pior   = int(linhas["ALVO_PIOR"])
        vetor_pior   = [int(x) for x in linhas["VETOR_PIOR"].split(",") if x.strip()]
        resultado[n] = {
            "n": n,
            "melhor": (vetor_melhor, alvo_melhor),
            "medio":  (vetor_medio,  alvo_medio),
            "pior":   (vetor_pior,   alvo_pior),
        }
    return resultado
 
 
# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------
 
def main():
    os.makedirs(DADOS_DIR, exist_ok=True)
 
    # Carrega entradas do C se disponível (só para os tamanhos do BT)
    entradas_c: dict[int, dict] = {}
    if os.path.exists(ENTRADAS_PATH):
        print(f"Lendo entradas compartilhadas de: {ENTRADAS_PATH}")
        entradas_c = parse_entradas(ENTRADAS_PATH)
    else:
        print("Arquivo de entradas compartilhadas não encontrado — gerando entradas próprias.")
 
    with open(SAIDA_CSV, "w", newline="") as f:
        writer = csv.writer(f)
        writer.writerow(["Algoritmo", "Cenario", "Tamanho_N",
                         "Tempo_Medio_ms", "Desvio_Padrao_ms"])
 
        print("\nIniciando bateria de testes...\n")
 
        # ── BACKTRACKING (tamanhos pequenos) ─────────────────────────────────
        print(">>> BACKTRACKING (O(2^n)) — tamanhos: " + str(TAMANHOS_BT))
        for n in TAMANHOS_BT:
            print(f"=== n = {n} ===")
            # usa entradas do C se existir para esse n; caso contrário gera novas
            bloco = entradas_c.get(n) or gerar_bloco(n)
            vetor_melhor, alvo_melhor = bloco["melhor"]
            vetor_medio,  alvo_medio  = bloco["medio"]
            vetor_pior,   alvo_pior   = bloco["pior"]
 
            testar_e_salvar("Backtracking", "Melhor Caso", n,
                            vetor_melhor, alvo_melhor, writer, eh_dp=False)
            testar_e_salvar("Backtracking", "Caso Medio",  n,
                            vetor_medio,  alvo_medio,  writer, eh_dp=False)
            testar_e_salvar("Backtracking", "Pior Caso",   n,
                            vetor_pior,   alvo_pior,   writer, eh_dp=False)
 
        # ── PROGRAMAÇÃO DINÂMICA (tamanhos maiores) ───────────────────────────
        print("\n>>> DP (O(n×W)) — tamanhos: " + str(TAMANHOS_DP))
        for n in TAMANHOS_DP:
            print(f"=== n = {n} ===")
            bloco = gerar_bloco(n)  # entradas do C não têm esses tamanhos
            vetor_melhor, alvo_melhor = bloco["melhor"]
            vetor_medio,  alvo_medio  = bloco["medio"]
            vetor_pior,   alvo_pior   = bloco["pior"]
 
            testar_e_salvar("DP", "Melhor Caso", n,
                            vetor_melhor, alvo_melhor, writer, eh_dp=True)
            testar_e_salvar("DP", "Caso Medio",  n,
                            vetor_medio,  alvo_medio,  writer, eh_dp=True)
            testar_e_salvar("DP", "Pior Caso",   n,
                            vetor_pior,   alvo_pior,   writer, eh_dp=True)
 
    print(f"\nTodos os testes finalizados! Resultados em {SAIDA_CSV}")
 
 
if __name__ == "__main__":
    main()